cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/ReplicaCapacityGoal.java

```diff
@@ -0,0 +1,261 @@
+      return false;
+    }
+    return true;
+  }
+
+  /**
+   * Update goal state after one round of self-healing / rebalance.
+   *
+   *  @param clusterModel The state of the cluster.
+   * @param excludedTopics The topics that should be excluded from the optimization proposal.
+   */
+  @Override
+  protected void updateGoalState(ClusterModel clusterModel, Set<String> excludedTopics)
+      throws AnalysisInputException, OptimizationFailureException {
+    // One pass is sufficient to satisfy or alert impossibility of this goal.
+    // Sanity check to confirm that the final distribution has less than the allowed number of replicas per broker.
+    ensureReplicaCapacitySatisfied(clusterModel);
+    // Sanity check: No self-healing eligible replica should remain at a decommissioned broker.
+    AnalyzerUtils.ensureNoReplicaOnDeadBrokers(clusterModel);
+    finish();
+  }
+
+  /**
+   * Sanity check: Ensure that the replica capacity per broker is not exceeded.
+   *
+   * @param clusterModel The cluster model.
+   */
+  private void ensureReplicaCapacitySatisfied(ClusterModel clusterModel) {
+    for (Broker broker : brokersToBalance(clusterModel)) {
+      int numBrokerReplicas = broker.replicas().size();
+      if (numBrokerReplicas > _balancingConstraint.maxReplicasPerBroker()) {
+        throw new IllegalStateException(String.format("Replicas in broker: %d exceeds the maximum allowed number of "
+            + "replicas: %d.", numBrokerReplicas, _balancingConstraint.maxReplicasPerBroker()));
+      }
+    }
+  }
+
+  /**
+   * Rebalance the given broker without violating the constraints of the current goal and optimized goals.
+   *  @param broker         Broker to be balanced.
+   * @param clusterModel   The state of the cluster.
+
+   * @param optimizedGoals Optimized goals.
+
+   * @param excludedTopics The topics that should be excluded from the optimization proposals.
+   */
+  @Override
+  protected void rebalanceForBroker(Broker broker, ClusterModel clusterModel, Set<Goal> optimizedGoals,
+      Set<String> excludedTopics) throws AnalysisInputException, ModelInputException, OptimizationFailureException {
+    LOG.debug("balancing broker {}, optimized goals = {}", broker, optimizedGoals);
+    for (Replica replica : new ArrayList<>(broker.replicas())) {
+      if (broker.isAlive() && broker.replicas().size() <= _balancingConstraint.maxReplicasPerBroker()) {
+        break;
+      }
+      if (excludedTopics.contains(replica.topicPartition().topic())) {
+        continue;
+      }
+
+      // The goal requirements are violated. Move replica to an available broker.
+      if (maybeApplyBalancingAction(clusterModel, replica, replicaCapacityEligibleBrokers(replica, clusterModel),
+          BalancingAction.REPLICA_MOVEMENT, optimizedGoals) == null) {
+        throw new OptimizationFailureException("Violated replica capacity requirement for broker with id " + broker.id() + ".");
```

Comment:
We may not want to throw exception if the current replica cannot be moved. It is possible that another replica can be moved away if we continue.